woocommerce/woocommerce
# Add the slack-test-report util

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).

### Changes proposed in this Pull Request:

Added the `slack-test-report` utility in `monorepo-utils` to replace the [automattic/action-test-results-to-slack](https://github.com/Automattic/action-test-results-to-slack) action which became unfit for our CI reporting needs after the latest changes. 
I decided on building our own tool instead of opening PRs on the action we were using, for speed, more flexibility and control.
The new tool borrows heavily from the mention action, but it's more lightweight, with less features.

Problems solved:
- sends messages only on failure (the previous action was also sending on skipped)
- can send a message for any job failure, with a custom job (or report) name. Example, send a message if the job which builds the matrix fails and no tests are running because of that.


### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

Here's a p1716564245022969-slack-C02180M7JFM notification for [this run](https://github.com/woocommerce/woocommerce/actions/runs/9226047800/). There are other notifications in that channel (you can ignore the missing PR titles or numbers, it's expected because I never set them because normally a notification is not sent to PRs).

Locally, you can set the required environment variables and run the tool with `pnpm utils slack-test-report -c failure`.
Use the `--help` option to see the accepted options.
Besides the program options there is a series of GitHub context environment variables that are expected for a complete message. See [the code](https://github.com/woocommerce/woocommerce/pull/47805/files#diff-79e4aa30d9db5c30554b77b1b0624b70c2fb342af2bb615693e8abec34c7480fR57) for the expected context. Missing context variables should not fail, but send an incomplete message.
